components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/jwt/JWTBearerGrantHandler.java

```diff
@@ -0,0 +1,653 @@
+                    assertion = param.getValue()[0];
+                    if (log.isDebugEnabled()){
+                        log.debug("JWT assertion found in request parameters: " + assertion);
+                    }
+                    break;
+                }
+            }
+
+            // check whether a non empty assertion string was acquired
+            if (assertion == null || StringUtils.isEmpty(assertion)) {
+                log.error("No Valid Assertion was found for " + OAUTH_JWT_BEARER_GRANT_TYPE);
+                return false;
+            }
+        }
+
+        //logging the JSON Web Token(JWT) assertion
+        if (log.isDebugEnabled()) {
+            log.debug("Received JSON Web Token assertion: " + assertion);
+        }
+
+        try {
+            signedJWT = SignedJWT.parse(assertion);
+
+            // logging the parsed JWT
+            if (log.isDebugEnabled()) {
+                logJWT(signedJWT);
+            }
+
+        } catch (ParseException e) {
+            // there's some error in the JSON Web Token
+            log.error("Error parsing the JWT received " + e.getMessage(), e);
+            return false;
+        }
+
+        // get the claims from the parsed JWT
+        ReadOnlyJWTClaimsSet claimsSet;
+        try {
+            claimsSet = signedJWT.getJWTClaimsSet();
+            if(log.isDebugEnabled()){
+                log.debug("Claim Set of the JWT: " + claimsSet.toJSONObject().toString());
+            }
+        } catch (ParseException e) {
+            log.error("Error when trying to retrieve claimsSet from the JWT", e);
+            return false;
+        }
+
+
+        /**
+         *  The JWT MUST contain an iss (issuer) claim that contains a unique identifier for the entity
+         *  that issued the JWT. In the absence of an application profile specifying otherwise, compliant
+         *  applications MUST compare Issuer values using the Simple String Comparison method
+         */
+        String jwtIssuer = claimsSet.getIssuer();
+
+        if (jwtIssuer == null || StringUtils.isEmpty(jwtIssuer)) {
+            log.debug("Issuer(iss) is empty in the JSON Web Token");
+            return false;
+        } else {
+
+            log.debug("Issuer Found in JWT: " + jwtIssuer);
+
+            // find and match a registered IDP to confirm that JWT is issued by a known IDP
+            try {
+
+                // TODO add an additional param called "issuer" in Oauth
+                /**
+                 // find a registered IDP with the 'Issuer' property matching that of the issuer value of JWT
+                 identityProvider = IdentityProviderManager.getInstance().getIdPByAuthenticatorPropertyValue(
+                 ISSUER,jwtIssuer, tenantDomain, false);
+                 **/
+
+                // TODO remove this after adding a property called "issuer" in Oauth Authenticator UI
+                identityProvider = createDummyIDP();
+
```

Comment:
Let's define an IdP with the name of the issuer and retrieve the IdP by the issuer.